### ISSUE TITLE
Error when decorator overrides constructor of class with static method 

### ISSUE THREAD

**Issue by Flybbit** [Microsoft/TypeScript#23704]
**TypeScript Version:**  2.9.0-dev.20180426 and previous

**Search Terms:** 
class decorator
override constructor

**Code**
```ts
function classDecorator() {
  return (constructor: new (...args: any[]) => any) => {
    return constructor;
  };
}

@classDecorator()
class Test {
    static run() { }
}


console.log(new Test());
```

**Expected behavior:**
Should be compiled with no error

**Actual behavior:**
error TS1238: Unable to resolve signature of class decorator when called as an expression.
  Type 'new (...args: any[]) => any' is not assignable to type 'typeof Test'.
    Property 'run' is missing in type 'new (...args: any[]) => any'.

It compiles just fine if I comment declaration of static method. Actually code works in playground despite compilation error. ts-none also executes this code, It seems to be false-positive error check, but it affects Angular compilation and breaks it.

[**Playground Link** ](http://www.typescriptlang.org/play/index.html#src=function%20classDecorator()%20%7B%0D%0A%20%20return%20(constructor%3A%20new%20(...args%3A%20any%5B%5D)%20%3D%3E%20any)%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20constructor%3B%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%40classDecorator()%0D%0Aclass%20Test%20%7B%0D%0A%20%20%20%20static%20run()%20%7B%20%7D%0D%0A%7D%0D%0A%0D%0A%0D%0Aconsole.log(new%20Test())%3B)

**Comment by mhegazy** [comment]
The decorator at runtime replaces the constructor function for the class. the compiler verifies that whatever it replaces it with has the same type. 

`Test.run` does not exist on the return type of your decorator function. define your decorator as a generic function instead:

```ts
function classDecorator() {
    return <T extends new(...args: any[]) =>any>(constructor: T) => {
        return constructor;
    };
}
```

**Comment by Flybbit** [comment]
Declaration as generic solves issue, thanks a lot. 

I thought compiler compares properties of passed and returned values and if I return passed parameter it should work. But it seems comparing returned value with type in declaration. That's why generics work as well as 
```ts
interface ctor {
  new(...args: any[]): any;
  run: any;
}

function classDecorator() {
  return (constructor: ctor ) => {
    return constructor;
    };
}

@classDecorator()
class Test {
    static run() { }
}
```
